examples/pytorch/pyg/train.py

```diff
@@ -0,0 +1,158 @@
+            their ordering to uniquely identify edges. (default: :obj:`None`)
+    """
+
+    def __init__(self, ge):
+        """Initialize DeepGNN graph store."""
+        super().__init__()
+        self.ge = ge
+
+    def _put_edge_index(self, edge_index, edge_attr) -> bool:
+        """To be implemented by :class:`GraphStore` subclasses."""
+        return False
+
+    def _remove_edge_index(self, edge_attr) -> bool:
+        """To be implemented by :class:`GraphStore` subclasses."""
+        return False
+
+    def _get_edge_index(self, edge_attr):
+        """To be implemented by :class:`GraphStore` subclasses."""
+        edge_type = int(edge_attr.edge_type[1])
+        edge = self.ge.sample_edges(
+            self.ge.edge_count(0), np.array(edge_type), SamplingStrategy.Random
+        )
+        edge_index = (
+            torch.Tensor(edge[:, 0]).long(),
+            torch.Tensor(edge[:, 1]).long(),
+        )
+        # assert False, edge_index
+        return edge_index
+
+    def get_all_edge_attrs(self):
+        """Obtain all edge attributes stored in the :class:`GraphStore`."""
+        output = []
+        # for i in range(self.ge.node_count(0)):
+        # node_type_0, edge_type, node_type_1
+        ta = EdgeAttr(("0", "0", "0"), "coo", size=[self.ge.node_count(0), 56944])
+        output.append(ta)
+
+        return output
+
+
+def train():
+    """Train the model."""
+    model.train()
+
+    total_loss = total_examples = 0
+    for data in tqdm.tqdm(loader):
+        data = data.to(device)
+        data = data[("0", "0", "0")]
+        optimizer.zero_grad()
+
+        h = model(data.x, data.edge_index)
+
+        h_src = h[data.edge_label_index[0]]
+        h_dst = h[data.edge_label_index[1]]
+        link_pred = (h_src * h_dst).sum(dim=-1)  # Inner product.
+
+        loss = F.binary_cross_entropy_with_logits(link_pred, data.edge_label)
+        loss.backward()
+        optimizer.step()
+
+        total_loss += float(loss) * link_pred.numel()
+        total_examples += link_pred.numel()
+
+    return total_loss / total_examples
+
+
+if __name__ == "__main__":
+    ge = PPI()
+    loader = LinkNeighborLoader(
+        (DeepGNNFeatureStore(ge), DeepGNNGraphStore(ge)),
+        batch_size=2048,
+        shuffle=True,
+        neg_sampling_ratio=1.0,
+        num_neighbors=[10, 10],
+        num_workers=6,
+        persistent_workers=True,
+        edge_label_index=(
+            ("0", "0", "0"),
+            torch.Tensor(ge.sample_edges(248388, np.array(0), SamplingStrategy.Random))
+            .long()[:, :2]
+            .T,
+        ),
+    )
+
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    model = GraphSAGE(
+        in_channels=50,
+        hidden_channels=64,
+        num_layers=2,
+        out_channels=64,
+    ).to(device)
+    optimizer = torch.optim.Adam(model.parameters(), lr=0.005)
+
+    for epoch in range(1, 6):
+        loss = train()
```

Comment:
Please, add a a test() method similar [to this](https://github.com/pyg-team/pytorch_geometric/blob/decec47e3c9cd67b8ddbd58a15edd00de0c530ef/examples/graph_sage_unsup.py#L57C1-L67C29)(with evaluation only on a test dataset) and an assert to make sure we get expected values from training.

Reply:
Is it possible with Cora since all edges have type 0? It would just be the training set accuracy which maybe is okay also?

Reply:
I added the test but even though loss is going down, test stays random, do you see anything wrong with it?

Reply:
Could you explain, what is random? f1 score stays low or non deterministic? If the former, you need to debug outputs of wrappers and compare them to relevant PyG tensors. If the latter, you need to set random seed. Either [globally](https://github.com/microsoft/DeepGNN/blob/1e35f511e6f18857e1aa0c774e9dee23b9adc88c/src/python/deepgnn/pytorch/common/utils.py#L31), or in the calls to sampling methods.